### PR TITLE
#4160 Add logging for user actions in multiplayer

### PR DESCRIPTION
**Pull request by delebota**
Logs most of the major user actions, let me know if I should add anything else to the log. I checklist of what is logged can be found at https://github.com/OpenRCT2/OpenRCT2/issues/4160#issuecomment-238757080

### CODE REVIEW

**Comment by IntelOrca** [comment]
You should reduce code duplication as much as possible.

**Comment by delebota** [comment]
Ok, the duplicate code for starting and appending the logs has been pulled out into new methods BeginLog and AppendLog, let me know if there is anything else to change.

Sorry for the merge, I should have made a separate branch in my repo but I forgot.

**Comment by IntelOrca** [comment]
Now you do need to rebase as there are merge conflicts, this should remove the previous merge.

**Comment by delebota** [comment]
I will take a look tonight, thanks!

**Comment by janisozaur** [comment]
Please also fix your commit messages https://github.com/OpenRCT2/OpenRCT2/wiki/Commit-Messages

File: src/game.c [hunk 1]
```diff
@@ -477,6 +477,132 @@ int game_do_command_p(int command, int *eax, int *ebx, int *ecx, int *edx, int *
...
+			char log_msg[256];
+			char* args[2] = {
+				(char *) player_name,
+				price_str
+			};
+
+			format_string(log_msg, STR_LOG_PARK_ENTRANCE_FEE, args);
+			network_append_server_log(log_msg);
+		}
+	}
```
Comment (IntelOrca) [inline src/game.c]: Please move this to its own function, otherwise game_command just becomes far too big.

**Comment by IntelOrca** [comment]
I think you need to log more actions, particularly removal and placement of scenery which are often the things done by grievers.

**Comment by delebota** [comment]
I will look into adding scenery commands in the next few days. Looks like I'll need to rebase again as well.

**Comment by janisozaur** [comment]
Doesn't seem to work. I connected as a client to a server, but no directory with server logs was created and even after I've manually created it, no logs were recorded.

**Comment by delebota** [comment]
It is setup to work only for the the server host and can only be enabled and disabled by them (similar to how 'block unknown players' is only an option for the host). It's meant as a tool to be able to go back and check for griefing so there isn't really any point to logging the data if you are a client.

**Comment by janisozaur** [comment]
1. In that case, why are there all the checks for client mode?
2. Given current landscape of openrct2 servers, it appears there are moderation teams put in place, they certainly would find this feature useful. I think enabling it for clients as well would be useful.

**Comment by ArcticX9** [comment]
As part of a moderation team on a popular server, it would be helpful for clients (such as the mod team) to be able to view this, as our server operator is not always available and leaves most of the server functions to the moderators aside from uploading new maps or restarting the server.

**Comment by telk5093** [comment]
I think is is required that server can know a client's hash (which is used in users.json) to make them spectator if they do terror unless there is no ban system yet. What about logging another one just after joining a new client to announce its hash?

**Comment by fromthere** [comment]
@telk5093 changing join log username(hash) has joined the game would be enough.

**Comment by delebota** [comment]
Should allow clients to log actions as well now!

**Comment by janisozaur** [comment]
@delebota can you rebase?

**Comment by delebota** [comment]
Should be good to go, assuming the CI checks go through ok.

**Comment by janisozaur** [comment]
Thanks, I'll try to review and test it soon.

File: src/localisation/string_ids.h [hunk 2]
```diff
@@ -3639,6 +3639,38 @@ enum {
 	STR_NO_LAND_RIGHTS_FOR_SALE_TIP = 5988,
 	STR_NO_CONSTRUCTION_RIGHTS_FOR_SALE_TIP = 5989,
 	STR_NO_LAND_OR_CONSTRUCTION_RIGHTS_FOR_SALE_TIP = 5990,
+    STR_LOG_SERVER_ACTIONS = 5991,
```
Comment (janisozaur) [inline src/localisation/string_ids.h]: This is a bit misaligned

File: src/localisation/string_ids.h [hunk 3]
```diff
@@ -3639,6 +3639,38 @@ enum {
...
+	STR_LOG_RIDE_PRICE = 6012,
+	STR_LOG_RIDE_SECONDARY_PRICE = 6013,
+	STR_LOG_PARK_NAME = 6014,
+	STR_LOG_PARK_CLOSED = 6015,
+	STR_LOG_PARK_OPEN = 6016,
+	STR_LOG_PARK_ENTRANCE_FEE = 6017,
+	STR_LOG_PLACE_SCENERY = 6018,
+	STR_LOG_REMOVE_SCENERY = 6019,
+	STR_LOG_EDIT_SCENERY = 6020,
+    STR_LOG_CLIENT_STARTED = 6021,
```
Comment (janisozaur) [inline src/localisation/string_ids.h]: misaligned

**Comment by IntelOrca** [comment]
@janisozaur It needs to flush the file after every written line though - otherwise you could end up losing log data if the game crashes.

**Comment by janisozaur** [comment]
Is it relevant, though? If a game has crashed, this log probably wouldn't have enough information to help us/host to debug and we aim to fix crashes anyway, for which we have dumps or stack traces.

I haven't looked yet into how exactly is this implemented, but e.g. "terraforming" (as named in "last player action" in "multiplayer" window) can cause hundreds of game commands being run in very short period of time, this could possibly cause same amount of file openings, writes, closures.

**Comment by IntelOrca** [comment]
> Is it relevant, though? If a game has crashed, this log probably wouldn't have enough information to help us/host to debug and we aim to fix crashes anyway, for which we have dumps or stack traces.

Without flushing, you can end up with quite large chunks of data not being written to the file in the event of a crash. And its also not just if it crashes, its that the user can't inspect the file and assume it contains the latest logged information.

**Comment by janisozaur** [comment]
I don't think files written to via syscalls are buffered all that long. Especially, if there's another reader connected. I believe a file can still be buffered but not flushed to FS when it's appended to (so a reader can connect `tailf` and just read the log as it appears), but when you close a file, SDL flushes it causing unnecessary I/O. I will have to confirm that's the case, though.

A bit kludgy, but a solution to that would be to flush it every, say, 5 seconds (if there's anything to log).

Also: in case of crash, everything was already `write()`-ten to, so we can just call `SDL_Close` on that FD to force the flush (which would still happen anyway, due to process closing and thus closing all owned handles which would trigger flushes on all buffers).

**Comment by janisozaur** [comment]
> Without flushing, you can end up with quite large chunks of data not being written to the file in the event of a crash.

Not true. The data has already been written to FD, your system handles the rest, even in case of crashing process.

https://linux.die.net/man/2/close

> A successful close does not guarantee that the data has been successfully saved to disk, as the kernel defers writes. It is not common for a file system to flush the buffers when the stream is closed. If you need to be sure that the data is physically stored use fsync(2). (It will depend on the disk hardware at this point.)

You can see for yourself in `SDL_rwops.c` there is no flush happening:

https://github.com/spurious/SDL-mirror/blob/d3f9978231577841304e160af50813c5e999c198/src/file/SDL_rwops.c

There is no mention of *flushing* things in `SDL_RWclose`, it only says this function should be used to "close and free an allocated SDL_RWops structure". I'm quite certain `close()` eventually comes, if not from the process itself, then on behalf of crashed process from the system. I can come up with a proof later on, if you want to.

Thus, I still insist on handling it the way I mentioned: open it whenever network state is changed and only close when disconnecting, without the penalty of opening and closing file on *each game command*.

**Comment by IntelOrca** [comment]
> Not true. The data has already been written to FD, your system handles the rest, even in case of crashing process.

As long as this is true for Windows as well then that's file and my point just stands at making sure that external applications can see the log updates without more than a second delay.

**Comment by janisozaur** [comment]
Sure, that's the intended use case for appending to a file. You can most likely get *less* delay if you forego `fopen()` and `fclose()`, as you reuse the buffer in memory. The data need not to be flushed to disk to be available to a reading process.

Even for Windows, [`CreateFile`](https://github.com/spurious/SDL-mirror/blob/d3f9978231577841304e160af50813c5e999c198/src/file/SDL_rwops.c#L106) is called without `FILE_FLAG_OVERLAPPED` flag set, which according to https://msdn.microsoft.com/en-us/library/windows/desktop/aa365683(v=vs.85).aspx#Synchronous_and_Asynchronous_I_O_Considerations implies synchronous access, so the whole discussion is pretty much moot.

@IntelOrca do you agree with me now?

**Comment by IntelOrca** [comment]
@janisozaur yes, I will test it when its done to see if I can see the changes in the file

Its just I have seen cases before when I open a log file and its only got half of a log line at the bottom of the file.

**Comment by janisozaur** [comment]
@delebota will you implement the requested change?

**Comment by janisozaur** [comment]
Creating or modifying the rides is not logged, at least creation should be marked.

**Comment by janisozaur** [comment]
A brief test on a live server shows it's a nice feature to have, but I find the information logged a bit lacking. Perhaps this is a nature of being a developer, but ideally I'd want to see finer grained details: location of changes (where applicable), state of cheats (on/off for ones you toggle), more actions logged (terraforming, peep picking, ride creation, …).

That said it is a good start and we can extend it in the future if need be.

**Comment by delebota** [comment]
Ride Creation should be getting logged, I will do some more testing though as there was quite a few conflicts in the last rebase that had to be resolved. I will also look into the buffer changes when I can.

**Comment by Gymnasiast** [comment]
Superseded by #5557.

**Pull request closed without merging**